src/compile.rs

```diff
@@ -682,6 +706,145 @@ fn check_required_config<'a>(
     ret.into_result(())
 }
 
+pub fn get_infos<'a>(
+    new_type_map: &TypeMap,
+    policies: &'a [PolicyFile],
+    classlist: &ClassList,
+    system_configs: &Option<&BTreeMap<String, &'a Argument>>,
+) -> Result<Vec<sexp::Sexp>, CascadeErrors> {
+    let ret = CascadeErrors::new();
+    // Get the policy files
+    let new_policies = get_policy_files(new_type_map, policies);
+
+    // Get the function infos
+    let mut new_func_map = get_funcs(&new_policies, new_type_map)?;
+
+    // Validate functions
+    let new_func_map_copy = new_func_map.clone(); // In order to read function info while mutating
+    validate_functions(
+        &mut new_func_map,
+        new_type_map,
+        classlist,
+        &new_func_map_copy,
+    )?;
+
+    // Get the policy rules
+    let new_policy_rules = get_policy_rules(&new_policies, new_type_map, classlist, &new_func_map)?;
+
+    let new_cil_tree = generate_sexp(
+        new_type_map,
+        classlist,
+        new_policy_rules,
+        &new_func_map,
+        system_configs,
+    )?;
+
+    ret.into_result(new_cil_tree)
+}
+
+pub fn get_reduced_types(
+    module: &ValidatedModule,
+    reduced_type_map: &mut TypeMap,
+    type_map: &TypeMap,
+    module_map: &ModuleMap,
+) -> Result<(), CascadeErrors> {
+    for t in &module.types {
+        if let Some(type_info) = type_map.get(t.name.as_ref()) {
+            if !reduced_type_map.iter().any(|(k, _v)| k == t.name.as_ref()) {
+                reduced_type_map.insert(t.name.to_string(), type_info.clone())?;
+            }
+        }
+        for parent in &t.inherits {
+            if let Some(parent_type_info) = type_map.get(parent.as_ref()) {
+                if !reduced_type_map.iter().any(|(k, _v)| k == parent.as_ref()) {
+                    reduced_type_map.insert(parent.to_string(), parent_type_info.clone())?;
+                }
+            }
+        }
+    }
+    for vm in &module.validated_modules {
+        if let Some(child_module) = module_map.get(vm.as_ref()) {
+            get_reduced_types(child_module, reduced_type_map, type_map, module_map)?;
+        }
+    }
+    Ok(())
+}
+
+pub fn get_policy_files<'a>(
```

Comment:
Okay, if I understand this right, the assumption here is that:

1. All the files we care about are declaration_files in types in our module.
2. Every function and rule in those files are needed.

Is that accurate?  I'm concerned about this assumption because:

1. A file may contain many types, only some of which are in the system.
2. A type may be "extend"ed using an extend block in some other file.

I think we need to rebrowse all files, but only descend into blocks of included types when generating the later data structure.